File: source/go/http_server.go
package main

import "fmt"

type Request struct {
	method string
	path   string
}

type Response struct {
	status int
	body   string
	method string
	path   string
}

func serverHandler(req Request) Response {
	var body string
	status := 200
	switch req.path {
	case "/":
		body = "Hello, World!"
	case "/about":
		body = "About Page"
	default:
		status = 404
		body = "Not Found"
	}
	return Response{status: status, body: body, method: req.method, path: req.path}
}

func main() {
	requestCh := make(chan Request, 10)
	responseCh := make(chan Response, 10)
	done := make(chan struct{})

	// Server goroutine
	go func() {
		for i := 0; i < 3; i++ {
			req := <-requestCh
			res := serverHandler(req)
			responseCh <- res
		}
	}()

	// Client goroutine
	go func() {
		requests := []Request{
			{method: "GET", path: "/"},
			{method: "GET", path: "/about"},
			{method: "GET", path: "/missing"},
		}
		for _, req := range requests {
			requestCh <- req
			res := <-responseCh
			fmt.Printf("[simulated] %s %s -> %d %s\n", res.method, res.path, res.status, res.body)
		}
		close(done)
	}()

	<-done
}

File: source/javascript/http_server.js
// Simulated HTTP server using request/response objects
async function handleRequest(req) {
    let body = "", status = 200;
    if (req.path === "/") body = "Hello, World!";
    else if (req.path === "/about") body = "About Page";
    else { status = 404; body = "Not Found"; }
    return { status, body, method: req.method, path: req.path };
}

async function main() {
    const requests = [
        { method: "GET", path: "/" },
        { method: "GET", path: "/about" },
        { method: "GET", path: "/missing" }
    ];

    for (const req of requests) {
        const res = await handleRequest(req);
        console.log(`[simulated] ${res.method} ${res.path} -> ${res.status} ${res.body}`);
    }
}

main();
